### DISCUSSION TITLE
codecov not working

### DISCUSSION THREAD

**Discussion by fabiocaccamo** [ymyzk/tox-gh-actions#170] [Q&A]
**Describe the bug**
I switched from Travis CI to GitHub actions for testing my django packages.
I'm using the same tox file, but now the coverage report is not uploaded to **codecov** anymore, as you can see [here](https://app.codecov.io/gh/fabiocaccamo/django-colorfield) there are no more reports uploaded since September 2021.

**To Reproduce**
 
* GitHub Actions configuration
This is the [workflow file](https://github.com/fabiocaccamo/django-colorfield/blob/master/.github/workflows/test-package.yml), it works pretty well (except the coverage reporting).

* tox and tox-gh-actions configuration:
This is my [tox file](https://github.com/fabiocaccamo/django-colorfield/blob/master/tox.ini)

* Link to the GitHub Actions job if a repository is public
https://github.com/fabiocaccamo/django-colorfield/actions/runs/1612327892

**Expected behavior**
The coverage report should be uploaded to [codecov](https://codecov.io/).

**Additional context**
By running tox locally, actually the coverage is 92%.

**Comment by ymyzk** [comment]
Hello. As far as I checked your logs quickly, I can see that the coverage itself it calculated on GitHub Actions but the codecov command is failing to upload. Could you check codecov's documentation and configure your repository correctly?

>   Error: Could not determine repo and owner
>  Tip: See all example repositories: https://github.com/codecov?query=example

**Comment by fabiocaccamo** [comment]
Hello @ymyzk,
`codecov` doesn't require any specific setting for open-source projects.

If I use the official codecov GitHub [action](https://github.com/marketplace/actions/codecov) directly with unittest it works, but if I use it running tests with tox it [fails](https://github.com/fabiocaccamo/python-fontbro/runs/4783122138?check_suite_focus=true).

**Comment by demberto** [comment]
The same thing is happening to me in my repo [tkinter-msgcat](https://github.com/demberto/tkinter-msgcat).

Basically what I do is run coverage commands inside tox, and change dir to `toxinidir`. Somehow `tox-gh-actions` doesn't understand this and the coverage reports don't get generated at the place they should be, which are later uploaded as artifacts and uploaded to codecov by a different job.

**Comment by fabiocaccamo** [comment]
@demberto maybe my solution works for you too:
- I stopped using tox inside my GitHub workflow
- I wrote the [fabiocaccamo/create-matrix-action](https://github.com/fabiocaccamo/create-matrix-action) for using a concise matrix syntax similar to the tox one *(if you will use it, star it!)*
- I started to use the [codecov/codecov-action](https://github.com/codecov/codecov-action)
- Everything works as expected 🚀

Here you can see a workflow of a python/django package of mine where I use: concise matrix + coverage + codecov: https://github.com/fabiocaccamo/django-admin-interface/blob/master/.github/workflows/test-package.yml

🍻

**Comment by demberto** [comment]
@fabiocaccamo thanks for trying to help. Looks like I need to stop using tox altogether in CI what a PITA. It seems author is not interested in this issue.

**Comment by ymyzk** [comment]
Hi all, I'm sorry for not being able to reply to this thread. I was actually busy with personal stuff lately.

I'm actually still not sure if this is a problem with tox-gh-actions. As tox-gh-actions doesn't do anything special which might affect codecov, this may be an issue with how to run tox and codecov on GitHub Actions using multiple jobs.

If someone is interested in digging into deeper on how tox + codecov + multiple GitHub Actions jobs works, I really appreciate it. Producing a minimal reproduction case is another way to help this. As this is an OSS without paid support, I cannot guarantee that all questions are answered.

Also, as the original author seems to have solved this issue in another approach. I'm also okay to close this issue for now.

**Reply by StevenLooman** [reply]
This does not seem to be a problem with `tox-gh-actions`. Tox still runs as expected. You can write the coverage report (XML) to a specific file using variables in `tox`. For example (change the `--cov` argument):

```ini
[tox]
envlist = py310, py311

[gh-actions]
python =
    3.10: py310
    3.11: py311

[testenv]
commands = py.test --cov=src --cov-report=term --cov-report=xml:coverage-{env_name}.xml {posargs}
```

This will result in two files, next to a report on the terminal:
* `coverage-py310.xml`
* `coverage-py311.xml`

Using [codecov/codecov-action](https://github.com/codecov/codecov-action), you can specify which file to upload:
```yml
    - name: Upload coverage reports to Codecov
      uses: codecov/codecov-action@v3
      if: ${{ hashFiles('coverage-py311.xml') != '' }}
      with:
        files: coverage-py311.xml
      env:
        CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
```

In this example only `coverage-py311.xml` is uploaded to [codecov.io](https://codecov.io). If you want to upload multiple files, you can do so easily by changing `files`  to `files: coverage-py310.xml, coverage-py311.xml` (as per their [arguments description](https://github.com/codecov/codecov-action#arguments)).

**Comment by ymyzk** [comment]
(I meant to click the `Comment` button instead of `Close with comment` button)
